packages/core/src/services/documents/utils.ts

```diff
@@ -29,3 +71,84 @@ export function existsAnotherDocumentWithSamePath({
 }) {
   return documents.find((d) => d.path === path) !== undefined
 }
+
+export async function resolveDocumentChanges({
+  originalDocuments,
+  newDocuments,
+}: {
+  originalDocuments: DocumentVersion[]
+  newDocuments: DocumentVersion[]
+}): Promise<{
+  documents: DocumentVersion[]
+  errors: Record<string, CompileError[]>
+}> {
+  const errors: Record<string, CompileError[]> = {}
+
+  const getDocumentContent = async (path: string): Promise<string> => {
+    const document = newDocuments.find((d) => d.path === path)
+    if (!document) {
+      throw new Error(`Document not found`)
+    }
+    return document.content
+  }
+
+  const newDocumentsWithUpdatedHash = await Promise.all(
+    newDocuments.map(async (d) => {
+      const metadata = await readMetadata({
+        prompt: d.content ?? '',
+        referenceFn: getDocumentContent,
+      })
+      if (metadata.errors.length > 0) {
+        errors[d.documentUuid] = metadata.errors
```

Comment:
What shape have these errors?

Reply:
They're instances of `CompileError`, which is an `Error` object with some extra stuff.